src/shims/fs.rs

```diff
@@ -548,6 +553,255 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.set_last_error(ebadf)?;
         Ok((-1).into())
     }
+
+
+    fn mkdir(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+        mode_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("mkdir")?;
+
+        #[cfg(target_os = "linux")]
+        let mode = this.read_scalar(mode_op)?.to_u32()?;
+        #[cfg(not(target_os = "linux"))]
+        let mode = this.read_scalar(mode_op)?.not_undef()?.to_u16()?;
+
+        let path = this.read_os_str_from_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+
+        let mut builder = DirBuilder::new();
+        #[cfg(target_family = "unix")]
+        {
+            use std::os::unix::fs::DirBuilderExt;
+            builder.mode(mode.into());
+        }
+        #[cfg(not(target_family = "unix"))]
+        let _mode = mode;
+        let result = builder.create(path).map(|_| 0i32);
+
+        this.try_unwrap_io_result(result)
+    }
+
+    fn rmdir(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("rmdir")?;
+
+        let path = this.read_os_str_from_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+
+        let result = remove_dir(path).map(|_| 0i32);
+
+        this.try_unwrap_io_result(result)
+    }
+
+    fn opendir(&mut self, name_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("opendir")?;
+
+        let name = this.read_os_str_from_c_str(this.read_scalar(name_op)?.not_undef()?)?;
+
+        let result = read_dir(name);
+
+        match result {
+            Ok(dir_iter) => {
+                let size = 1;
+                let kind = MiriMemoryKind::Env;
```

Comment:
That... looks like you are leaking memory. Why that?

Reply:
This is explained in the new doc comment I just pushed above -- this allocation will be deallocated if/when the interpreted program calls `closedir` on the directory stream.

Reply:
`Env` is to be used by the env var emulation and other things where it is okay for memory not to be deallocated. It is not the right choice here -- not closing the directory stream clearly should count as a memory leak.

If no other memory kind matches, please just add a new one.

Reply:
On second thought, it seems like you don't actually ever write anything to that memory? I don't understand why you are even allocating anything then... this strikes me as very strange, and the comments do not explain it.

Why don't you just use a unique ID in a similar way to what we do with file descriptors?

Reply:
OK, the newest commit gets rid of the allocations and tracks everything with IDs. I originally passed around pointers to otherwise-unused allocations to line up with the FFI interface using pointers, but it looks like stuffing IDs where the `DIR*` pointers were works just fine.